File: lintcode/construct-binary-tree-from-inorder-and-postorder-traversal.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/construct-binary-tree-from-inorder-and-postorder-traversal/

"""
Definition of TreeNode:
"""
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None

class Solution:
    def helper_func(self, inorder, reverse_postorder):
        if len(inorder) == 0:
            return None

        root = TreeNode(reverse_postorder[0])
        for i in xrange(len(inorder)):
            if inorder[i] == reverse_postorder[0]:
                break

        inorder_left = inorder[:i]
        inorder_right = inorder[i+1:]
        reverse_postorder_left = reverse_postorder[len(reverse_postorder)-i:]
        reverse_postorder_right = reverse_postorder[1:len(reverse_postorder)-i]

        root.left = self.helper_func(inorder_left, reverse_postorder_left)
        root.right = self.helper_func(inorder_right, reverse_postorder_right)

        return root

    """
    @param inorder : A list of integers that inorder traversal of a tree
    @param postorder : A list of integers that postorder traversal of a tree
    @return : Root of a tree
    """
    def buildTree(self, inorder, postorder):
        # write your code here
        return self.helper_func(inorder, postorder[::-1])

if __name__ == '__main__':
    s = Solution()
    root = s.buildTree([1,2], [2,1])
    print root.val,root.right.val

File: leetcode-2/construct-binary-tree-from-inorder-and-postorder-traversal.cc
// Question: https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return helper(inorder, postorder, 0, (int)inorder.size()-1, 0, (int)postorder.size()-1);
    }

    TreeNode* helper(vector<int>& inorder, vector<int>& postorder, int il, int ir, int pl, int pr) {
        if (pl > pr) {
            return NULL;
        }
        if (pl == pr) {
            return new TreeNode(postorder[pr]);
        }
        int i = il;
        while (i <= ir) {
            if (inorder[i] == postorder[pr]) {
                break;
            }
            ++i;
        }
        int len = i - il;
        auto root = new TreeNode(postorder[pr]);
        root->left = helper(inorder, postorder, il, i-1, pl, pl+len-1);
        root->right = helper(inorder, postorder, i+1, ir, pl+len, pr-1);
        return root;
    }
};
